src/main/java/vg/civcraft/mc/citadel/listener/InventoryListener.java

```diff
@@ -1,50 +1,90 @@
 package vg.civcraft.mc.citadel.listener;
 
+import org.bukkit.Location;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
 import org.bukkit.block.Container;
+import org.bukkit.block.Dispenser;
+import org.bukkit.block.DoubleChest;
+import org.bukkit.block.Dropper;
+import org.bukkit.block.Hopper;
+import org.bukkit.block.data.Directional;
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.EventPriority;
 import org.bukkit.event.Listener;
 import org.bukkit.event.inventory.InventoryMoveItemEvent;
 import org.bukkit.inventory.Inventory;
-
+import org.bukkit.inventory.InventoryHolder;
 import vg.civcraft.mc.citadel.ReinforcementLogic;
 import vg.civcraft.mc.citadel.model.Reinforcement;
+import vg.civcraft.mc.civmodcore.api.BlockAPI;
+import vg.civcraft.mc.civmodcore.util.Iteration;
 
 public class InventoryListener implements Listener {
 
-	// prevent sucking reinforcements out of reinforced containers with hoppers on
-	// different groups or filling into them
 	@EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
 	public void onInventoryMoveItemEvent(InventoryMoveItemEvent event) {
-		Inventory sourceInv = event.getSource();
-		Reinforcement sourceRein = null;
-		if (sourceInv.getHolder() instanceof Container) {
-			sourceRein = ReinforcementLogic.getReinforcementProtecting(((Container) sourceInv.getHolder()).getBlock());
+		Inventory fromInventory = event.getSource();
+		Inventory destInventory = event.getDestination();
+		InventoryHolder fromHolder = fromInventory.getHolder();
+		InventoryHolder destHolder = destInventory.getHolder();
+		// Determine the reinforcement of the source
+		Reinforcement fromReinforcement = null;
+		if (fromHolder instanceof Container) {
+			Container container = (Container) fromHolder;
+			fromReinforcement = ReinforcementLogic.getReinforcementProtecting(container.getBlock());
+		}
+		else if (fromHolder instanceof DoubleChest) {
+			DoubleChest doubleChest = (DoubleChest) fromHolder;
+			Block doubleChestBlock = doubleChest.getLocation().getBlock();
+			Location chestLocation = doubleChestBlock.getLocation(); // Yes this is necessary
+			Location otherLocation = BlockAPI.getOtherDoubleChestBlock(doubleChestBlock).getLocation();
```

Comment:
Is it really neccessary to be this specific in regards to hoppers? For example FactoryMods pipes will call InventoryMoveItemEvents to be compatible with other plugins, but those will move from arbitrary containers into other arbitrary containers. I'm also not really a fan of the relative location get here, there should be a safer way to do that via the source/dest holder.

Reply:
> there should be a safer way to do that via the source/dest holder.

Unfortunately not, inventory locations are finicky and don't necessarily match the [expected location](https://imgur.com/Zy9EV7x) when dealing with double chests, which makes resolving which reinforcement to check against more difficult. I have added a check into `/ctf` so you can't double reinforce blocks (including double chests), *however* this is just a check for new reinforcements and so doesn't do anything about the three years of activity on the server that has likely resulted in thousands of double reinforced double chests. And those chests might be reinforced to different groups on each side.. so the hopper really does need to know exactly what block it's draining from.. and there's sadly no way to do that with the native event.

Reply:
Double reinforced double chests are a feature that we do not intend to remove. Mismatched reinforcements on different halves of a double chest are also intended and good.

The bug there is that CTR can not do that, not that CTF can do it.

I'll agree on the relative location check under these conditions though.

Reply:
> Double reinforced double chests are a feature that we do not intend to remove.

Oh, okay never mind I'll remove that then :P but yeah the relative check is a bit hacky but it's only applied to Hoppers, so factory pipes shouldn't be subject to that.. unless you're factory piping from a double chest and into a hopper?